File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        sort(arr.begin(), arr.end());
        map<long long, long long> mpans;
        long long res = 0;
        long mod = 1e9+7;
        for (int i = 0; i < arr.size(); i++){
            mpans[arr[i]] = 1;
        }
        
        for(int i = 1; i < arr.size(); i++){
            int x = arr[i];
            int t = sqrt(x);
            long long ans = 1;
            for(int j = 2; j <= t; j++){
                if (x % j == 0){
                    int p = x / j;
                    if (p == j){
                        ans += (mpans[j]%mod * mpans[j]%mod) %mod;
                    }
                    else{
                        ans += (((mpans[j]%mod * mpans[p]%mod)%mod) * 2)%mod;
                    }
                }
            }
            mpans[x] = ans % mod;
        }
        for(auto it: mpans){
            res += it.second;
        }
        return res%mod;
    }
};

File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        mod = 10**9+7
        arr.sort()
        res = 0
        mpans = {}
        
        for i in range(len(arr)):
            mpans[arr[i]] = 1
        
        for i in range(1, len(arr)):
            x = arr[i]
            t = int(x**0.5)
            ans = 1
            for j in range(2, t+1):
                if x % j == 0:
                    p = x//j
                    if (j in mpans) and (p in mpans):
                        if p == j: 
                            ans += mpans[j] * mpans[j]
                        else:
                            ans += (mpans[j] * mpans[p])*2
            mpans[x] = ans % mod
        for i in mpans:
            res += mpans[i]
        return res % mod
                            